File: Data Structure/Lab 5/Anagram.c
//  Design anagram game using array.  
//  Allow a user to enter N words and store it in an array.  
//  Generate a random number between 0 to N-1.  
//  Based on the random number generated display the word stored at that 
// index of an array and allow user to enter its anagram.  
//  Check whether the word entered by the user is an anagram of displayed 
// number or not and display an appropriate message.  
//  Given a word A and word B. B is said to be an anagram of A if and only if the 
// characters present in B is same as characters present in A, irrespective of 
// their sequence. For ex: “LISTEN” == “SILENT” 

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <time.h>
#include <ctype.h>

int main() {
    int n, i;

    printf("Enter no. of words: ");
    if (scanf("%d", &n) != 1 || n <= 0) {
        printf("Invalid number of words.\n");
        return 1;
    }

    char words[n][100];
    printf("Enter %d words in small letters:\n", n);
    for(i = 0; i < n; i++) {
        scanf("%s", words[i]);
    }

    srand(time(NULL));

    int score = 0;       // Initialize score to 0
    char choice;         // To check if user wants to play again

    // Loop to allow multiple rounds of play
    do {
        int random = rand() % n;
        printf("\nThe word is: %s\n", words[random]);

        char user[100];
        printf("Enter Anagram of the given word: ");
        scanf("%s", user);

        int isAnagram = 1; 

        if(strlen(user) != strlen(words[random])) {
            isAnagram = 0; 
        } else {
            int freq1[26] = {0};
            int freq2[26] = {0};

            for(i = 0; i < strlen(words[random]); i++) {
                char temp = words[random][i];

                int index = temp - 'a';

                freq1[index]++;
            }

            for(i = 0; i < strlen(user); i++) {
                char temp = user[i];

                int index = temp - 'a';

                freq2[index]++;
            }

            for(i = 0; i < 26; i++) {
                if(freq1[i] != freq2[i]) {
                    isAnagram = 0;
                    break;
                }
            }
        }

        // Update score based on the outcome
        if(isAnagram) {
            score += 10; // Add 10 points for a correct answer
            printf("Hurray! Words are Anagrams.\n");
        } else {
            printf("Oh no! Words are NOT Anagrams.\n");
        }

        // Display current score
        printf("Your Current Score: %d\n", score);

        printf("\nDo you want to play another round? (y/n): ");
        scanf(" %c", &choice); // Notice the space before %c to clear newline character

    } while(choice == 'y' || choice == 'Y');

    // Final game over screen
    printf("\n--- GAME OVER ---\n");
    printf("Your Final Score is: %d\n", score);

    return 0;
}

File: Data Structure(Java)/Lab 4/Anagram.java
//  Design anagram game using array.  
//  Allow a user to enter N words and store it in an array.  
//  Generate a random number between 0 to N-1.  
//  Based on the random number generated display the word stored at that 
// index of an array and allow user to enter its anagram.  
//  Check whether the word entered by the user is an anagram of displayed 
// number or not and display an appropriate message.  
//  Given a word A and word B. B is said to be an anagram of A if and only if the 
// characters present in B is same as characters present in A, irrespective of 
// their sequence. For ex: “LISTEN” == “SILENT"

import java.util.Arrays;
import java.util.Scanner;

public class Anagram {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter no. of words:");
        int n = sc.nextInt();

        if(n <= 0){
            System.out.println("Invalid number !");
            sc.close();
            return;
        }

        String[] words = new String[n];

        System.out.println("Enter "+n+ "words in small letters :");
        for(int i = 0 ; i < n ; i++){
            words[i] = sc.next();
        }

        int score = 0;
        char choice;

        do{
            int random = (int)(Math.random() * n);

            String word = words[random];

            System.out.println("Enter Anagram for :"+word);

            System.out.print("Enter Anagram :");
            String[] user = new String[100];

            String userguess = sc.next();

            boolean isAnagram = false;

            for(int i = 0 ; i < userguess.length() ; i++){
                user[i] = String.valueOf(userguess.charAt(i));
            }

            if(words[random].length() == userguess.length()){

                char[] wordc = word.toCharArray();
                char[] userg = userguess.toCharArray();

                Arrays.sort(wordc);
                Arrays.sort(userg);

                    if(Arrays.equals(wordc, userg)){
                        isAnagram = true;
                    }
            }

            if(isAnagram){
                System.out.println("Hurray! Words are Anagram.");
                score +=10;
            }
            else{
                System.out.println("Oops!! Words are not Anagram.");
            }

            System.out.println("Your current score:"+score);
            System.out.println("Do you want to play another round(y/n):");
            choice = sc.next().toLowerCase().charAt(0);

        }while(choice == 'y');

        System.out.println("Final score :"+score);

        sc.close();
    }
}
